### ISSUE TITLE
Powrmatic

### ISSUE THREAD

**Issue by real-bombinho** [danielrivard/innova-controls#41]
Hello and apologies for hijacking, I do not expect the problem to be rooted in this repo.

I do have a Powrmatic Vision Compact, which is basically the very same device as the Innova 2.0 Mini, just rebranded.

Since using the WiFi automation I have noticed that the device completely disappears from WiFi for hours or even days (comes back on after power down). This makes remote controlling very flaky, has someone else noticed such behaviour?

side note:
I had a brief look at the ESP-01 fitted to the control board. The (lean) firmware seems not to be Arduino based, which is not much of a surprise. I used the exploration to get a firmware backup.

**Comment by danielrivard** [comment]
Yeah that firmware is not very stable, when we are too aggressive on the polling it crashes and needs a restart. This is why I originally set the scan interval to 10 minutes in the home assistant integration.

Now it is configurable by the user. On my setup, I have it set at 3 minutes between polls and it seems to be doing ok. But when I started working on this library, I think 30 seconds would make it crash from time to time.

**Comment by real-bombinho** [comment]
I did not (yet) thank you for your work, a big Thank You.

Occasionally I also do set the time of the device on "/api/v/1/set/time" with something like "tz=Europe/London&value=1730307624" <- being UTC Linux time stamp. Albeit its RTC seems to revert to 0 infrequently, when without access to the internet. 
It does store the timezone in the flash. I hope it counterchecks for changes rather writing senselessly.

**Comment by HossainGh** [comment]
Hi @real-bombinho 
I have an Innova 2.0 unit with the same issue. The firmware is 1.0.42. Could you let me know if you found a solution to fix the problem or upgrade the firmware?
I have an additional ESP-01 at home. Can I copy the software to the new ESP-01 and resolve the issue? (However, I don't know how I can do it 😅)

**Comment by real-bombinho** [comment]
Hello @HossainGh ,

unfortunately I have not spent as much time as I would love to on this issue. It gets far worse. The ESP01 goes completely AWOL at times, sometimes for days, rendering the WiFi inaccessible so control via WiFi is rather patchy. I already have considered to use the hard wired remote switch capability inside the device (wire bridge) to make reliably sure that the device is switched off when I transition between battery and grid.

I think, the best way forward would be to analyse the communication between the controller and the ESP01 and replace it with some FOSS. There has AFAIK never been an update/upgrade for the ESP firmware despite it being rather unreliable and having serious safety issues.

**Comment by HossainGh** [comment]
Hi,

I have done some hardware experiments that might be interesting. I disassembled the control panel for more investigation. The panel's main brain is the R5F513T3ADFJ microcontroller. https://mou.sr/42NaVCJ
Also, the ESP8266 used on ESP01s board has a better antenna circuit. I believe ESP8266 is used only as a web server and a Wi-Fi chip on this device. It's worth mentioning that GPIO pins that are not used and not connected to any traces. It seems the microcontroller communicates with ESP01s board by TX and RX pins (what is the protocol? I leave it for experts!). 

I had another ESP01s board with a different memory chip manufacturer. I was able to clone ESP01s and install the new one (cloned board) on the panel. Unfortunately, I observed the same behavior and wifi connectivity issue! 

I connected the original ESP01s to a power supply (3.3V) and left it aside. Since I have already set up with Wi-Fi credentials, it connected to my wifi. I monitored the original ESP01s with ping and the status API. It worked for several hours without any issues! I assume something on the panel board is interfering with the ESP chip and causing the connectivity issue.

I will keep you posted with new findings :)

**Comment by real-bombinho** [comment]
Thank you @HossainGh, currently I am a bit short for time, so you have saved me some experimenting this way.

Did you identify the pins in use for comms?

**Comment by HossainGh** [comment]
I am happy the information is helpful to you, @real-bombinho . Unfortunately, I didn't pay attention to the communication pins of the microcontroller chip. Since the PCB has a coating, it was shiny and I didn't take a picture either. I only noticed TX and RX pins of ESP connected to the borad. Non of GPIO pins were used in this device.
I have bin file of ESP.  I attached the file which can be useful for you. (change ext from txt to bin)

[innova_rst.txt](https://github.com/user-attachments/files/20121697/innova_rst.txt)

**Comment by HossainGh** [comment]
Guys, I did something crazy... 😅

So I got SO annoyed with the WiFi dropouts that I literally bought a logic analyzer, reverse-engineered the entire Modbus protocol (I learned for this project!), and wrote new firmware from scratch (As my first ESPHome YAML!).

**I cannot survive with issues like this!** 😂

Good news: It works perfectly! Stable WiFi, Home Assistant integration, the whole package.

🔗 https://github.com/HossainGh/esphome-innova

Everything's documented - firmware, register map, installation guide, HA config. Just replace your ESP-01S and you're good to go!

Hope this saves someone else the pain of disappearing air conditioners! 🙌

**Comment by real-bombinho** [comment]
@HossainGh , this is brilliant.

I had an early Christmas present this year myself (a logic analyser) and was hoping to do what you have done already now. A big Thank You ahead for your work. The lack of a reliable WiFi has become a real pain in the last months, with the Powrmatic more often than not being accessible. Whatever festivity you celebrate, have a brilliant time.

**Comment by danielrivard** [comment]
This is really nice work @HossainGh! I'm not sure I'm willing to do the modification on my unit just yet, but this is pretty good.

I wonder if your ESPHome config could expose an API similar to the original one of the Innova. This way, we could reuse the integration I built without the need for the Template Climate. But still your work is very good!

Also I wonder if the new units that changed the communication protocol are also using an ESP-01S, and if it could support the same kind of switch and provide a local API instead of their new cloud only API.
